x-pack/filebeat/input/awscloudwatch/input.go

```diff
@@ -0,0 +1,360 @@
+			}
+			in.logger.Error("getLogEventsFromCloudWatch failed: ", err)
+			continue
+		}
+
+		in.logger.Debugf("sleeping for %v before checking new logs", in.config.ScanFrequency)
+		time.Sleep(in.config.ScanFrequency)
+		in.logger.Debug("done sleeping")
+	}
+}
+
+func parseARN(logGroupARN string) (string, string, error) {
+	arnParsed, err := arn.Parse(logGroupARN)
+	if err != nil {
+		return "", "", errors.Errorf("error Parse arn %s: %v", logGroupARN, err)
+	}
+
+	if strings.Contains(arnParsed.Resource, ":") {
+		resourceARNSplit := strings.Split(arnParsed.Resource, ":")
+		if len(resourceARNSplit) >= 2 && resourceARNSplit[0] == "log-group" {
+			return resourceARNSplit[1], arnParsed.Region, nil
+		}
+	}
+	return "", "", errors.Errorf("cannot get log group name from log group ARN: %s", logGroupARN)
+}
+
+func (in *awsCloudWatchInput) getLogEventsFromCloudWatch(svc cloudwatchlogsiface.ClientAPI) error {
+	errC := make(chan error, 1)
+	cwCtx := &cwContext{
+		refs: 1,
+		errC: errC,
+	}
+	defer cwCtx.done()
+
+	ctx, cancelFn := context.WithTimeout(in.inputCtx, in.config.APITimeout)
+	defer cancelFn()
+
+	i := 0
+	nextToken := ""
+	currentTime := time.Now()
+	startTime, endTime := getStartPosition(in.config.StartPosition, currentTime, in.prevEndTime, in.config.ScanFrequency)
+	in.logger.Debugf("start_position = %s, startTime = %v, endTime = %v", in.config.StartPosition, time.Unix(startTime/1000, 0), time.Unix(endTime/1000, 0))
+
+	// overwrite prevEndTime using new endTime
+	in.prevEndTime = endTime
+
+	for nextToken != "" || i == 0 {
+		// construct FilterLogEventsInput
+		filterLogEventsInput := in.constructFilterLogEventsInput(startTime, endTime, i, nextToken)
+
+		// make API request
+		req := svc.FilterLogEventsRequest(filterLogEventsInput)
+		resp, err := req.Send(ctx)
+		if err != nil {
+			in.logger.Error("failed FilterLogEventsRequest", err)
+			return err
+		}
+
+		// get token for next API call
+		nextToken = ""
+		if resp.NextToken != nil {
+			nextToken = *resp.NextToken
+		}
+
+		logEvents := resp.Events
+		in.logger.Debugf("Processing #%v events", len(logEvents))
+
+		err = in.processLogEvents(logEvents, cwCtx)
+		if err != nil {
+			err = errors.Wrap(err, "processLogEvents failed")
+			in.logger.Error(err)
+			cancelFn()
+		}
+
+		// increase counter after making FilterLogEventsRequest API call
+		i++
+	}
+	return nil
+}
+
+func (in *awsCloudWatchInput) constructFilterLogEventsInput(startTime int64, endTime int64, i int, nextToken string) *cloudwatchlogs.FilterLogEventsInput {
+	filterLogEventsInput := &cloudwatchlogs.FilterLogEventsInput{
+		LogGroupName: awssdk.String(in.config.LogGroup),
+		StartTime:    awssdk.Int64(startTime),
+		EndTime:      awssdk.Int64(endTime),
+	}
+
+	if len(in.config.LogStreams) > 0 {
+		filterLogEventsInput.LogStreamNames = in.config.LogStreams
+	}
+
+	if in.config.LogStreamPrefix != "" {
+		filterLogEventsInput.LogStreamNamePrefix = awssdk.String(in.config.LogStreamPrefix)
+	}
+
+	if i != 0 {
```

Comment:
this could just be `nextToken != ""` and avoid passing `i`

Reply:
haha yes I think `i` was doing something else and I forgot to remove it 😂 Thanks!!